numba/stencil.py

```diff
@@ -0,0 +1,617 @@
+                if isinstance(index, tuple) or isinstance(index, list):
+                    for i in range(len(index)):
+                        te = index[i]
+                        if isinstance(te, ir.Var) and te.name in const_dict:
+                            te = const_dict[te.name]
+                        if isinstance(te, int):
+                            neighborhood[i][0] = min(neighborhood[i][0], te)
+                            neighborhood[i][1] = max(neighborhood[i][1], te)
+                        else:
+                            raise ValueError("Non-constant used as stencil index.")
+                    index_len = len(index)
+                elif isinstance(index, int):
+                    neighborhood[0][0] = min(neighborhood[0][0], index)
+                    neighborhood[0][1] = max(neighborhood[0][1], index)
+                    index_len = 1
+                else:
+                    raise ValueError("Non-tuple or non-integer used as stencil index.")
+                if index_len != ndim:
+                    raise ValueError("Stencil index does not match array dimensionality.")
+
+        return (neighborhood, relatively_indexed)
+
+
+    def get_return_type(self, argtys):
+        if config.DEBUG_ARRAY_OPT == 1:
+            print("get_return_type", argtys)
+            ir_utils.dump_blocks(self.kernel_ir.blocks)
+
+        if not isinstance(argtys[0], types.npytypes.Array):
+            raise ValueError("The first argument to a stencil kernel must be the primary input array.")
+
+        typemap, return_type, calltypes = compiler.type_inference_stage(
+                self._typingctx,
+                self.kernel_ir,
+                argtys,
+                None,
+                {})
+        if isinstance(return_type, types.npytypes.Array):
+            raise ValueError("Stencil kernel must return a scalar and not a numpy array.")
+
+        real_ret = types.npytypes.Array(return_type, argtys[0].ndim,
+                                                     argtys[0].layout)
+        return (real_ret, typemap, calltypes)
+
+    def _install_type(self, typingctx):
+        """Constructs and installs a typing class for a StencilFunc object in
+        the input typing context.  If no typing context is given, then
+        _install_type() installs into the typing context of the
+        dispatcher object (should be same default context used by
+        jit() and njit()).
+        """
+        _ty_cls = type('StencilFuncTyping_' +
+                       str(hex(self.id).replace("-", "_")),
+                       (AbstractTemplate,),
+                       dict(key=self, generic=self._type_me))
+        typingctx.insert_user_function(self, _ty_cls)
+
+    def compile_for_argtys(self, argtys, kwtys, return_type, sigret):
+        # look in the type cache to find if result array is passed
+        result = None
+        for (_argtyps, _sig, _res, _typemap, _calltypes) in self._type_cache:
```

Comment:
Can be simplified to `(sig, res, typemap, calltypes) = self._type_cache[argtys]` 
See https://github.com/numba/numba/pull/2457/files#r148394639